### ISSUE TITLE
Utilize HTML5 Clipboard API when feasible

### ISSUE THREAD

**Issue by JamesMGreene** [zeroclipboard/zeroclipboard#171]
Utilize the HTML5 Clipboard API when feasible [in addition to aligning event names with it via #105].

The API is currently being standardized @ http://www.w3.org/TR/clipboard-apis/.

The working group may even be receptive to hearing suggestions from us as well, if we have any to offer. Personally, I'm still trying to figure out what the heck that spec enables us to do (other than paste-capturing, which is not helpful for ZeroClipboard).

**Comment by codylerum** [comment]
Chrome Launch Issue: https://code.google.com/p/chromium/issues/detail?id=437908

**Comment by JamesMGreene** [comment]
Thanks for the link, @codylerum!

The biggest problem is still that inability to feature detect Clipboard API support, which makes determining the need for Flash awkwardly late (the first click would potentially be wasted by trying to detect if the Clipboard API is supported). :sad:

**Comment by JamesMGreene** [comment]
When the HTML Clipboard API is discoverable and usable in 1+ browsers, we should revisit any issues that have the "HTML TopLayer" label assigned.

**Comment by jab** [comment]
Landed in Chrome 43:
https://code.google.com/p/chromium/issues/detail?id=437908#c8
http://updates.html5rocks.com/2015/04/cut-and-copy-commands

**Comment by jonrohan** [comment]
Here's what github does to enable copy to clipboard without flash. It's based on useragent sniffing and not feature detection, which is why it's not in ZC yet. https://gist.github.com/jonrohan/81085b119d16cdd7868a

**Comment by JamesMGreene** [comment]
Thanks, @jonrohan!

**Comment by dandv** [comment]
:+1:

BTW, http://stackoverflow.com/questions/26336138/how-can-i-copy-to-clipboard-in-html5-without-using-flash could use some input from you, @JamesMGreene :)

**Comment by silviapfeiffer** [comment]
Using the HTML Clipboard API if it's available or when Flash is disabled would really make a difference to us. The github approach seems to make sense. Is this in development?

**Comment by tomsommer** [comment]
What is the GitHub approach?

If the Clipboard API is availiable, I see no reason at all to use Flash.

**Comment by motti** [comment]
This is getting more urgent as Mozilla appears to be heading towards blocking flash by default and other browsers may follow soon. [Firefox makes click-to-activate Flash the default (mozilla.org) | Hacker News](https://news.ycombinator.com/item?id=9883246)

**Comment by silviapfeiffer** [comment]
GitHub approach see link shared by @jonrohan above. They discover the web api via browser version check.

**Comment by JamesMGreene** [comment]
Integrating the HTML Clipboard API into ZeroClipboard is not under active development yet but I may be able to squeeze in time for some long overdue bug fixes and a few releases next week. Here's hoping, anyway! :+1:

**Comment by JamesMGreene** [comment]
Since Firefox began blocking (or forcing "Ask to Activate" status) on all previous versions of Flash Player [other than the very latest release] as of July, and Chrome, IE10+, and Firefox are all capable of doing basic clipboard injection natively now (supposedly), this enhancement needs to be shifted to a pretty immediate top priority.

**Comment by arxpoetica** [comment]
Ported comment from #606 (“Will the Chrome flash-block announcement have any impact on ZeroClipboard”):

> See: https://plus.google.com/+GoogleAds/posts/2PmwKinJ7nj for more info
> Also: https://news.ycombinator.com/item?id=10133771 (sub discussion: https://news.ycombinator.com/item?id=10133915)

(Links have details on announcement.)

**Comment by tomsommer** [comment]
Just copy what GitHub does, it makes sense to use the native API when possible and then flash as a fallback.

**Comment by JamesMGreene** [comment]
I'll be integrating the native API in as soon as I can but feel free to implement something yourself in the meanwhile if you need a stop-gap solution.

**Comment by plessbd** [comment]
Fo anyone trying to use the native api this might be helpful, they even give props to zeroclipboard
https://hacks.mozilla.org/2015/09/flash-free-clipboard-for-the-web/

**Comment by cpeterso** [comment]
One Firefox gotcha highlighted in the article:

In Firefox 40 and earlier, `document.queryCommandSupported("copy"/"cut")` returns true but calling `document.execCommand("copy"/"cut")` will throw a SecurityException. The copy/cut commands were implemented, but restricted to privileged code like add-ons. Checking for this exception is necessary to safely feature-detect support for `document.execCommand("copy"/"cut")` in Firefox.

**Comment by fbender** [comment]
> One Firefox gotcha highlighted in the article:

And clarify further (e.g. for those who were initially as confused as I was): 

1. Firefox 41 (will be released mid-September) will support copy/cut via `execCommand`.
2. Firefox 40 (now on release) and previous versions throw when `document.execCommand("copy")` is used.
3. `document.queryCommandSupported("copy")` always returns true (at least in Firefox), regardless of if the call to `execCommand` will throw or work.

**Comment by mzgol** [comment]
To add to @fbender's list, starting from Firefox 41:

1. `document.queryCommandSupported('paste')` will fail (https://bugzilla.mozilla.org/show_bug.cgi?id=1161721).
2. `document.queryCommandEnabled('cut'/'copy')` will return `true` only when inside a user-generated event handler (https://bugzilla.mozilla.org/show_bug.cgi?id=1162952#c9).

**Comment by JamesMGreene** [comment]
So in adding the HTML Clipboard API as the preferred underlying clipboard mechanism in ZeroClipboard, the few things to keep in mind are:
 - Prefer feature detection (e.g. in Firefox)
 - Be careful about cross-browser feature detection (e.g. as @cpeterso commented above, Firefox used to throw an `Error`
 - Use browser detection when feature detection is limited (e.g. in Chrome, we'll have to actually inspect the Chromium version due to their feature detection bug https://code.google.com/p/chromium/issues/detail?id=476508 )
 - I would like to add a new configuration value (e.g. `ZeroClipboard.config({ allowSecurityPrompts: true })`, default value === `false`) in ZeroClipboard to give ZeroClipboard consumers the control to opt-out of using the HTML Clipboard API in IE/Edge if they don't want their end users to be presented with a security permissions prompt/dialog.
     - I'm hopeful Edge will remove that security prompt in the future if they have not already done so (I haven't tested that yet).
     - Additionally, it would be a reasonable question to consider if the the IE/Edge Clipboard API should still be attempted (regardless of the configuration value) if Flash is not a viable clipboard injection mechanism for a given end user's IE/Edge instance.
     - I'm allow unsure if there is any good way to feature detect this state, or if it will also need to become an unfortunate bit of IE-specific browser detection code.

**Comment by hallvors** [comment]
Some experiments in feature detection: http://jsbin.com/butaxaziyo/edit?js,output

queryCommandEnabled() only returning true inside event handlers is intended to help discover the "works in certain trusted event handlers" support. This may not be a great idea, given that it's inconsistent with return values for all other commands. The return value from execCommand() should also give a relatively good indication of whether the command worked (returns false for failures).

I assume the IE devs want to kill the prompt *if* the non-prompt model is proven safe enough by the other implementors and experience. I'd suggest not adding IE-specific hacks to prefer Flash.

**Comment by hallvors** [comment]
(It's of course rather easy to detect the false positive feature detection in Fx < 40 - 

    var supported = document.queryCommandSupported('copy');
    try{document.execCommand('copy')}catch(e){supported = false}

should do the trick. Only problem is if you happen to *have* privileges you'll actually cause a random copy command..)

**Comment by Yaffle** [comment]
@hallvors , do `document.queryCommandEnabled("copy");` instead of `document.execCommand('copy')` and you will not cause a copy command.

**Comment by hallvors** [comment]
@Yaffle - thanks!

I think feature detection with queryCommandSupported() and queryCommandEnabled() is going to work - with one gotcha being that in Blink, 'cut' is only considered 'enabled' if there is a selection *inside* an editable element *and* you're on a user-initiated thread. However, I think it's fine to just use the 'copy' command when detecting support and assume 'cut' support is equivalent (to avoid having to create editable elements in the DOM and mess with focus just for feature detection).

**Comment by JamesMGreene** [comment]
@Yafile @hallvors Does `document.queryCommandEnabled("copy")` throw an Error in Firefox < 41, then?

If not, this won't suffice either as that _should_ actually return `false` unless tested during a user-initiated event.

**Comment by JamesMGreene** [comment]
> I assume the IE devs want to kill the prompt if the non-prompt model is proven safe enough by the other implementors and experience.

I'd certainly hope so! :confused:

> I'd suggest not adding IE-specific hacks to prefer Flash.

I think the difference in behavior is too important to _not_ allow for handling it differently via configuration.

**Comment by JamesMGreene** [comment]
@jdalton @BenjamP: Can either of you help us find a Microsoft Edge contact for discussing Microsoft's intentions for the Clipboard API since @BenjamP left the company?

**Comment by jdalton** [comment]
Yep, can do.

**Comment by jdalton** [comment]
FWIW the current official status is listed as [under consideration](https://dev.modern.ie/platform/status/clipboardapi/?filter=f3f0000bf&search=clipboard). I know we mark interop issues like this at a higher priority. As to where it falls in the current rank of feature work I can't speak to that.

**Comment by JamesMGreene** [comment]
@jdalton @jonathansampson Is there a particular developer/engineer who took up the mantle of Editing/Selection/Clipboard APIs in @BenjamP's wake?

**Comment by jdalton** [comment]
What specifically are you wanting to discuss?

**Comment by hallvors** [comment]
@jdalton: the official status might as well be "partially implemented" given that the spec is partially based on what IE is already doing :)
I suppose what we really would like to know (and I know it's something you most likely won't be able to tell us with certainty) is whether and by when IE will switch to the spec's security model instead of the ask-the-user-with-a-modal-prompt model you're using now.
(If you worry about back-compat you could go for a combined solution, where you accept copy/cut from user-initiated threads w/o prompting like the spec says and other browsers are aligning on - but also have the prompt/domain white-listing model as a fallback for other cases..)

**Comment by hallvors** [comment]
@JamesMGreene I've been looking through some of ZC's code and wow, there's a bit more code than I expected ;) so my first question is: what would be your view on making the first script the browser loads a simple "loader" that picks Flash or HTML5 and inserts a SCRIPT element loading the desired version? We'd likely send significantly less code to clipboard-API compatible browsers.

(One counter-argument is that you do make RTF quite an important citizen - and I'm not sure what browser vendor's view on permitting JS to write RTF is and will be. If we have both implementations we could experiment with switching to using Flash if the site uses RTF..but it's tricky to get that right, I guess..)

**Comment by JamesMGreene** [comment]
@jdalton: As @hallvors mentioned, we would indeed mostly be seeking information of Edge timelines for (a) changing the security model and (b) supporting more data formats than `"Text"` (`"text/plain"` in the newer implementations) and `"URL"` (something like `"text/uri-list"` in newer implementations), particularly `"text/html"` and maybe `"application/rtf"`.

Also, I've also seen a number of claims that the clipboard functionality was "partially implemented"/available as of IE10 (e.g. in the related [HTML5 Rocks article](http://updates.html5rocks.com/2015/04/cut-and-copy-commands)) but I know this functionality has been in IE since IE5 and I'm not aware of any actual changes made in IE10. Can you comment on that?

**Comment by JamesMGreene** [comment]
@hallvors:

> I've been looking through some of ZC's code and wow, there's a bit more code than I expected ;) so my first question is: what would be your view on making the first script the browser loads a simple "loader" that picks Flash or HTML5 and inserts a SCRIPT element loading the desired version? We'd likely send significantly less code to clipboard-API compatible browsers.

Not a bad idea!  I'd be down for going that route so long as we also provide some pre-built complete versions as well.

> (One counter-argument is that you do make RTF quite an important citizen - and I'm not sure what browser vendor's view on permitting JS to write RTF is and will be. If we have both implementations we could experiment with switching to using Flash if the site uses RTF..but it's tricky to get that right, I guess.)

I don't consider RTF a critical functionality for the most part but it is something we've supported for quite a while now just because Flash 10+ has always been able to support injecting it into the clipboard.

**Comment by hallvors** [comment]
I really doubt that RTF has much of a future here - main reasons
* the importance of mobile - where we also want clipboard functionality but RTF is certainly not a first-class citizen. I'd expect it to be fully unsupported, but I don't have data to back that up..
* the sane reluctance of browser vendors when facing features with unpredictable security implications.
* it's not a web "citizen" in the first place.

So, my suggestion for an application/rtf "roadmap" is:
* Call it deprecated. Output a warning in the console if ZC is asked to process RTF. Start doing that ASAP.
* Once we get as far as making HTML5 default, add a pref for preferring Flash when available. Authors who depend on writing RTF to the clipboard must opt-in to using Flash over HTML5.

**Comment by nhducit** [comment]
I try to make demo to use zeroClipboard and document.execCommang()

https://github.com/nhducit/nh-copy

```js
(function () {
    //if the SWF file is hosted in different directory as ZeroClipboard.js, you need to set the URL
    //should do this when minify source code
    //ZeroClipboard.config( { swfPath: 'bower_components/zeroclipboard/dist/ZeroClipboard.swf' } );

    var copyEmailBtn = document.querySelector('.js-emailcopybtn');

    //Chrome bug: Calling queryCommandSupported() for cut or copy always returns false until after a user interaction.
    //use getChromeVersion to check if current browser is chrome and chrome version ? 43
    //use flash by default because jsCopy function may cause error
    if (isFlashInstalled()) {
        try {
            zeroCopy();
        } catch (e) {
            jsCopy()
        }

    } else if (!!document.queryCommandSupported('copy') || getChromeVersion() >= 43 || getFirefoxVersion() >= 41) {
        //select button element
        //TODO: use modernizr to detect flash
        //add copy click listener
        copyEmailBtn.addEventListener('click', jsCopy);
    } else {
        alert('Your browser need to update or have to install flash to use copy button');
    }
    /**
     * fallback function use flash
     */
    function zeroCopy() {
        //if there is error occur use zeroClipboard instead
        //currently we can not trigger click on element to make zeroClipboard work due to flash security

        //https://github.com/zeroclipboard/zeroclipboard/issues/109
        //https://github.com/zeroclipboard/zeroclipboard/pull/110
        //http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/System.html#setClipboard()
        var zeroClipboard = new ZeroClipboard(document.querySelector('.js-emailcopybtn'));
        zeroClipboard.on('copy', function (event) {
            console.log('use zeroClipboard');
            var clipboard = event.clipboardData;
            clipboard.setData('text/plain', document.querySelector('.js-emaillink').innerHTML);
        });
    }

    /**
     * js copy without flash
     */
    function jsCopy(/*event*/) {
        console.log('use javascript');
        // Select the email link anchor text
        var emailLink = document.querySelector('.js-emaillink');
        //The Range interface represents a fragment of a document that can contain nodes and parts of text nodes.
        var range = document.createRange();
        //The Range.selectNode() method sets the Range to contain the Node and its contents.
        //The parent Node of the start and end of the Range will be the same as the parent of the referenceNode.
        range.selectNode(emailLink);
        //use selection API select selected range
        var selection = window.getSelection();
        selection.addRange(range);

        try {
            //throw new Error('blah');
            // Now that we've selected the anchor text, execute the copy command
            var successful = document.execCommand('copy');
            var msg = successful ? 'successful' : 'unsuccessful';
            console.log('Copy email command was ' + msg);
        } catch (err) {
            alert('Copy have problem select text and copy manually');
        }

        // Remove the selections
        if (selection.removeRange && typeof selection.removeRange === 'function') {
            // removeRange(range) when it is supported
            selection.removeRange(range);
        } else {
            selection.removeAllRanges();
        }
        //known issue
        //if clipboard already have data, chrome show an error message 'Discontiguous selection is not supported'
        //BUT it work :)
    }

    /**
     * detect flash
     * @returns {boolean|*}
     */
    function isFlashInstalled() {
        var ie_flash;
        try {
            ie_flash = (window.ActiveXObject && (new ActiveXObject('ShockwaveFlash.ShockwaveFlash')) !== false)
        } catch (err) {
            ie_flash = false;
        }
        return ((typeof navigator.plugins != 'undefined' && typeof navigator.plugins['Shockwave Flash'] == 'object') || ie_flash);
    }

    /**
     * get chrome version
     * @returns {boolean|number} chrome version or false is current browser is not chrome
     */
    function getChromeVersion() {
        var raw = navigator.userAgent.match(/Chrom(e|ium)\/([0-9]+)\./);
        return raw ? parseInt(raw[2], 10) : false;
    }

    function getFirefoxVersion() {
        var raw = navigator.userAgent.match(/Firefox\/([0-9]+)\./);
        return raw ? parseInt(raw[1], 10) : false;
    }
})();
```

**Comment by paulirish** [comment]
Folks probably saw this, but today Zeno published https://github.com/zenorocha/clipboard.js 

The browser support is as follows 
<a href="https://cloud.githubusercontent.com/assets/39191/10150543/f0368468-65f4-11e5-8fea-bcd94bea5e8b.png"><img src="https://cloud.githubusercontent.com/assets/39191/10150543/f0368468-65f4-11e5-8fea-bcd94bea5e8b.png" width=250></a>

**Comment by karlhorky** [comment]
Thanks for the heads up @paulirish. Looks like a nice option.

**Comment by JamesMGreene** [comment]
Thanks for sharing, @paulirish.  I see @zenorocha must've liked a few of our ideas since he used them in clipboard.js. :+1:

It's a good light option but I know there will definitely still be a need to make ZeroClipboard support HTML5 but fallback to Flash for many consumers out there.

**Comment by zenorocha** [comment]
Actually I haven't seen ZeroClipboard's code when writing clipboard.js. My initial idea was to write a Web Component that extends `<button>` and not a standalone library, but when I realized there was few people using `execCommand` I decided to give it a try :)

Although there's some momentum going on with clipboard.js I'm pretty sure most people will continue to use ZeroClipboard. It's a stable project and it has better browser support.

Looking forward to see your HTML5 implementation :)

**Comment by JamesMGreene** [comment]
@zenorocha: 

> Actually I haven't seen ZeroClipboard's code when writing clipboard.js.

I figured you had (or at least looked at the documentation) since you also use the attributes `data-clipboard-target` and `data-clipboard-text`.  Either way, whatever. :smile:



> Although there's some momentum going on with clipboard.js

That's an understatement, my friend.  Your repo has rocketed from 0 to almost as many stars (~4500) as ZeroClipboard's (~5000) in a period of mere days.  Quite an impressive feat, IMHO. :clap:

**Comment by tomsommer** [comment]
Time to get it implemented then, gogo.

**Comment by hallvors** [comment]
I'm looking into it right now.

**Comment by hallvors** [comment]
Implementation notes/suggestions:
* One global click listener on document
* Detects whether the target element of a click matches any of the following:
    * Has a data-clipboard-text attribute
    * Has a data-clipboard-target attribute
    * Has been passed to the ZeroClipboard clip() method
    * Has been passed to the ZeroClipboard constructor
* If a matching element is found, the click listener will do the following:
    * Prepare data (question: which of data-clipboard-target, data-clipboard-text and setData()&friends takes priority?)
    * Fire ZC's copy event (question: is this expected to only fire at the "client" related to the specific element? We don't really need a client/core model in the HTML5 version.. I'd prefer to do only the minimum we need for back-compat)
    * Add a temporary copy event listener to the document
    * Trigger that listener with ```document.execCommand('copy')``` and use setData()
    * Remove that listener
    * Fire ZC's aftercopy event and forget data

Question: do you think it matters to have mouseover/out listeners to set .zeroclipboard-is-hover type class names? I'd hope we can get away with killing those, but if we must it's not a lot of code and it's written already.

I'm planning to change the build so that ZeroClipboard.js is a very tiny loader, just feature-detects HTML5 clipboard and adds a SCRIPT element loading either ZeroClipboard-HTML5.js or ZeroClipboard-Flash.js.

**Comment by fbender** [comment]
> I'm planning to change the build so that ZeroClipboard.js is a very tiny loader, just feature-detects HTML5 clipboard and adds a SCRIPT element loading either ZeroClipboard-HTML5.js or ZeroClipboard-Flash.js.

For speed and simplicity, I suggest bundling `ZeroClipboard-HTML5.js` with `ZeroClipboard.js` since the source should be rather tiny (hitting the network for that is probably overkill), and only load `ZeroClipboard-Flash.js` on-demand. The need for loading the extra file will diminish over time as more and more browsers get updated, and they won't need to load anything extra after the initial file.

**Comment by JamesMGreene** [comment]
I would agree with @fbender's point about including the HTML5 code by default, though we'll have to verify the size of it.

@hallvors: I'll try to get back to your proposed flow and question(s) today.

**Comment by hallvors** [comment]
@JamesMGreene OK, I've started mucking around and plan to check in some work-in-progress code shortly, maybe this evening.

**Comment by gked** [comment]
Heya folks,
My name is Grisha and I was told you were looking for an MS Edge representative on Editing side.
From what I've read so far, there were couple  of questions(please correct me if I misunderstood):

	1. Edge to kill security prompts in copy/cut execCommands - MS Edge doesn't have any security prompts for execCommands copy/cut.
	2. Edge to support Clipboard API - currently, we support clipboardData.
I can't speak of deadlines but  Clipboard API is definitely on our radar.

**Comment by hallvors** [comment]
The hardest part of this is not writing code - very little should need to be added. The hard part is that some of the API (both internal and external) no longer makes sense, and I'm not sure how much we have to care about back compat and how things typically were used. So..

## Notes on API for ZC HTML5

### Core:

ZeroClipboard constructor 
 - consider always returning the ZeroClipboard object itself, not create instances (though if we keep "client.destroy" we may run into problems with pages that intend to destroy only some instance..?)

config
state
 - kept

isFlashUnusable 
 - outputs warning in console, returns null. Or throws?

on
off
handlers
emit
 - kept only for back-compat, simplified if possible. emit is fixed to return a simpler object for 'copy'

create
 - not sure. Warning in console. Could return the ZC object itself?

destroy
 - removes event listeners from document for sure (click) (but what if there are more instances? ZeroClipboard.destroy() aways killed everything?)

setData
 - keep almost as is but throw or warn for RTF
getData
clearData

focus / activate
blur / deactivate
activeElement
 - not sure. Console warning and no-op?

### Client

_createClient
 - remember any element(s) passed to it (?)

on
off
handlers
emit
 - ??

clip
 - remember any element(s) passed to it

unclip
 - forget element(s) passed to it

elements
 - report what elements are remembered

destroy
 - ??
 - I think this should be removed or be a noop.

setText
setHtml
 - will be kept as-is (just calling the global setData anyway)
setData
 - will end up throwing for RTF
setRichText
 - ends up always throwing

clearData
getData

 - All of these set*, get* and clear* no longer check for a valid instance to throw if there isn't one. There's only one global data store and no concept of "instances" of ZC.

### Internal stuff
Mouseover/out event handlers, classes to indicate hovered state etc
 - remove all

**Comment by hallvors** [comment]
(I'm not sure if the event listeners were typically used by web content or more of an internal thing. If it's more of an internal thing, could we get away with dropping it?)

**Comment by zenorocha** [comment]
Hey guys, I wrote a [polyfill for document.queryCommandSupported](https://github.com/zenorocha/document.queryCommandSupported).

I didn't want to include it on clipboard.js because we could collaborate on this separately ;)

**Comment by groovenectar** [comment]
Can anyone find some cases where this logic would fail in anything but extremely marginal or very intentional cases? http://jsfiddle.net/mmy8yj57/2/

Here are the tests I've run...

Returns `friendlyId` value (non-false) in:

* Chrome 47 Dev, Chrome 45, Chrome 43, Android 4.4 Chrome, Android 4.2 Chrome
* MS Edge, IE11, IE10, Windows Phone 8.1
* FF43, FF41, FF, Android 4.4 FF 41
* Opera 32, Opera 29

Returns `false` in: Safari 5.1, Chrome 42, IE9, FF40, Opera 28

So all the tests that I've run so far are returning the expected values (note that this is a browser test and not an `execCommand()` test, since that seems to be the way to go for this right now). I have seen the GitHub copy to clipboard working in Android 4.4 Chrome and FF, but am not sure if WP8.1 is a false positive (probably good to go there too). Supports arbitrary version numbers thanks to a post by @LeJared, or could use parseInt() to keep it to major versions without the extra helper function.

Here are the helper functions:

	// http://stackoverflow.com/a/16187766
	var cmpVersions = function (a, b) {
		var i, l, d;

		a = a.split('.');
		b = b.split('.');
		l = Math.min(a.length, b.length);

		for (i=0; i<l; i++) {
			d = parseInt(a[i], 10) - parseInt(b[i], 10);
			if (d !== 0) {
				return d;
			}
		}

		return a.length - b.length;
	};

	var nativeClipboardSupport = function() {
		var browserSupport = [
			{
				friendlyId: 'firefox',
				vendorMatch: 'Firefox',
				minVersion: 41
			},
			{
				friendlyId: 'opera',
				vendorMatch: 'OPR',
				minVersion: 29
			},
			{
				friendlyId: 'ms',
				vendorMatch: '(?:MSIE|IE|Edge|Trident.+rv)',
				minVersion: 10
			},
			{ // Chrome last b/c included in others sometimes
				friendlyId: 'chrome',
				vendorMatch: 'Chrome',
				minVersion: 43
			}
		];

		for (var i = 0; i < browserSupport.length; i++) {
			var regex = browserSupport[i].vendorMatch + '(?:/| |:)([\\d\\.]+)';
			var matches = navigator.userAgent.match(new RegExp(regex));
			if (matches) {
				if (cmpVersions(matches[1], browserSupport[i].minVersion.toString()) > 0) {
					return browserSupport[i].friendlyId;
				}
			}
		}
	
		return false;
	};

**Comment by hallvors** [comment]
For the record, my "work in progress" detection code which isn't truly well-tested yet is this:

```
var _html5ClipboardSupported = false;
  try{
    /* In some browsers, in particular Firefox < 40, queryCommandSupported() will
     * return true because the command is "supported" in scripts with extra privileges
     * - but trying to use the API will throw. We use two functions below, but the order
     * matters: if queryCommandEnabled() throws, we will not use queryCommandSupported().
     * queryCommandEnabled() is expected to return false when not called from a
     * user-triggered thread, so it's only called here to see if it throws..
     */
    _html5ClipboardSupported = document.queryCommandEnabled("copy") || document.queryCommandSupported("copy");
  }catch(e){}
```

**Comment by groovenectar** [comment]
@hallvors, nice concise start to ideal actual feature detection.

This may be of interest: https://developers.google.com/web/updates/2015/04/cut-and-copy-commands

Particularly these parts:

> The difference between document.queryCommandSupported() and document.queryCommandEnabled() is that cut and copy could be supported by a browser, but if no text is currently selected, they won’t be enabled.

A solution to that could be programmatically selecting some text before checking.

> Calling queryCommandSupported() for cut or copy always returns false until after a user interaction. This prevents you from disabling your UI for browsers which don’t actually support it.

I'm not sure how that is in different browsers and I'm not sure if could have been intentional / by design. It is mentioned in this article as a known issue / bug report.

As mentioned in the thread, `execCommand('copy')` can only be used in a user-defined callback. I seemed to be having trouble with `queryCommandEnabled()` and/or `queryCommandSupported()` returning inconsistent results. I would run a simple test, get true, return the same test, get false, with no clear pattern that I could recognize.

Something like this seemed to produce more consistent results (I adapted your `or` condition because it's better than what I had):

	var hasNativeClip;
	$(window).on('cliptest', function() {
		try {
			hasNativeClip = document.queryCommandEnabled("copy") || document.queryCommandSupported("copy");
		} catch (e) {
			hasNativeClip = false;
		}
	}).trigger('cliptest').off('cliptest');

It was in jQuery and I don't have an example without jQuery at the moment but you get the idea...

Here is another related article that had apparently been hiding pretty well in previous research attempts: https://hacks.mozilla.org/2015/09/flash-free-clipboard-for-the-web/

**Comment by hallvors** [comment]
If queryCommandSupported() is reliable enough (and the false positive in older Fx can be weeded out by the exception from queryCommandEnabled()), we don't need to fake selections, and we don't need to worry about queryCommandEnabled() having its intricate and slightly idiosyncratic requirements about selections and user-triggered threads. (That's likely why you got inconsistent results..might that be the case?)

I'm *hoping* it will work to 

a) Use my snippet above for feature detection
b) Optionally use queryCommandEnabled() only in the very thread you would like to trigger copy from as an extra verification that the native API is supported. (And yes, for some browsers that means making sure there is a selection - although document.execCommand('copy') should work without as long as you set data from a copy event listener..).

**Comment by groovenectar** [comment]
That sounds good.. I do think that is queryCommandEnabled requires a selection for whatever reason, it's probably worth creating a selection to avoid a potentially needless check. Plus if there's a click, any existing user selection I think should be considered lost(?). A lot of our most common browsers are evergreen, but I'm wondering if there will end up being any benefit to a browser detection fallback for certain edge cases with this particular anomoly.

**Comment by hallvors** [comment]
This is the current state of my experiments:
https://gist.github.com/hallvors/07f0f1a79a042098641b
Note: Chrome will (for now) fall back to Flash given this code, but they are fixing a queryCommandSupported() bug now so it should work shortly.
Second note: this is barely tested. Review and feedback very welcome - just comment in the gist..

**Comment by hallvors** [comment]
(If you'd like to test in Chrome already, the polyfill above will help - thanks @zenorocha  :))

**Comment by zenorocha** [comment]
You're welcome @hallvors! And you're right, Chrome 48 already fixed this problem.

**Comment by tomsommer** [comment]
So is this happening or not?

**Comment by miketaylr** [comment]
> @hallvors: I'll try to get back to your proposed flow and question(s) today.

Do you have any time in the near future to respond to @hallvors' proposal, @JamesMGreene?

**Comment by hallvors** [comment]
See the current state here:
https://github.com/hallvors/zeroclipboard/tree/useclipboardapi
(many changes - one big checkin, sorry about that..)

Here's sort of a roadmap:
- [x] Make the build system build two separate implementations
- [ ] Adjust code so that the now obsolete parts are not included in the "HTML5" version (this is almost finished, I think)
- [x] Th e"HTML5" version loads by default, feature / API detecting to embed Flash version if required.  (I'm not sure if this code is solid enough)
- [ ] Check with some other code found by searching for "ZeroClipboard" on GitHub if the new ZC is a drop-in replacement (as it's intended to be)
- [ ]  Figure out the test story. I expect some tests will fail with the "HTML5" version, so we probably need to mark them as relevant for the Flash version only..

**Comment by hallvors** [comment]
@jonrohan I'm not sure which parts of ZC you've been working on earlier, but would you have time to review and comment some of this new code?

**Comment by jonrohan** [comment]
> I'm not sure which parts of ZC you've been working on earlier, but would you have time to review and comment some of this new code?

I'll do my best, but ZC has been re-thought so much, that i don't have a full mental picture of how the pieces of the API are built together anymore. @JamesMGreene is the current mastermind, which leaves a pretty terrible bus-factor. :worried:

**Comment by hallvors** [comment]
I tried to split the work into slightly more manageable commits in https://github.com/hallvors/zeroclipboard/tree/html5ify but it's still probably a lot of stuff to read. Some bug fixing commits too, so not a super-organised fake history ;)
Functionality-wise I think it's getting there..

**Comment by mrazzari** [comment]
@hallvors I went ahead and tested you html5ify branch as a drop-in replacement, and it didn't work.
Then I tried building from source, same thing but with different errors. 
Makes me think I'm doing something wrong. 

I realize this is little more than a +1, but I'd love to get this working, as the Flash fallback option still makes a lot of sense "in the real world".

**Comment by cpeterso** [comment]
@hallvors, what is the next action to move this feature forward? Do you just need review feedback from @JamesMGreene?

**Comment by hallvors** [comment]
@cpeterso The main missing piece (except figuring out why it didn't yet work for @mrazzari ) is figuring out a way to test sanely. The tests probably need to be modal (somehow) and run either the Flash tests or the HTML5 tests depending on what is supported. This makes me a bit uneasy though - the test setup becomes somewhat complicated and it's not going to be clear what assertations run.. So I'm not sure what the best approach is.

**Comment by hallvors** [comment]
I've done a little bit of work on making the tests modal. The build should finish without stopping on test errors once that lands on https://github.com/hallvors/zeroclipboard/tree/html5ify (not sure why it didn't show up yet - GitHub trouble?)

**Comment by mrazzari** [comment]
@hallvors I'm happy to try and repro if you point me to a working demo URL, to make sure it's not a build or config error on my end.

**Comment by hallvors** [comment]
@mrazzari my html5ify branch now builds fine here. Does your project happen to be public? I'll admit upfront I don't have a sufficient understanding of how people tend to use ZC in their code.

**Comment by mrazzari** [comment]
I'm using it pretty much OOTB:

```
jQuery(document).ready(function($){
   var client = new ZeroClipboard( $("#saveButton"));
});

<input value="xxx" type="text" readonly="readonly" id="secretLink">
<input id="saveButton" type="button" name="saveButton" value="Save &amp; Copy" data-clipboard-target="secretLink" />
```

But just to be clear, my earlier tests were done building your branch from source, and using a local Jekyll build of the zeroclipboard.org demo pages.

Sadly I didn't note the errors I was getting back then. I'll give it another go. But to rule out the possibility of a build error on my end, it may be useful if you had a public working demo.

**Comment by hallvors** [comment]
I was gaping over way too much in those initial attempts - trying to change the way the scripts are loaded and yank out all that was SWF-related for the native clipboard API version. It was hard to untangle correctly and probably even harder for @JamesMGreene and others to review. Mea culpa, it was not a good way to start this work.

Here's a different approach - this is basically "what's the smallest change we can make to ensure ZeroClipboard doesn't use Flash if there's clipboard API support"?
https://github.com/hallvors/zeroclipboard/tree/minimalNativeClipboardImpl
I've been testing this against this page:
http://hallvord.com/temp/moz/zeroclipboard/test.html
(not that it's *well* tested so far.. Test results from a mix of browsers and platforms welcome. Also, if there are common use cases that page doesn't have - or if you find bugs in it - please let me know.)

**Comment by tomsommer** [comment]
@hallvors This is exactly the approach we need. Go for it.

**Comment by hallvors** [comment]
I'm hoping this is ready to merge (but it would sure be nice if more people tested the URL above in various browsers and reported their findings here! Hint, hint!)

https://github.com/zeroclipboard/zeroclipboard/pull/650

**Comment by groovenectar** [comment]
Works on Chrome/Nexus5X

**Comment by peteallen** [comment]
Tests 2, 3, and 4 fail in Safari 9.1.1 on OS X 10.11.5. All tests pass in Chrome 50.0.2661.102.

**Comment by hallvors** [comment]
@peteallen thanks for testing! Looked into the Safari issue - the script does not attempt to use the native API, so apparently Safari's implementation isn't quite up for it yet (I think this is correct). I've added some error handling to the test case so it will print out a warning if Flash is disabled and it can't use a native implementation.
(This also is useful to check in other browsers that it emits no Flash error if it *does* use the native API - and we do indeed have a bug here, since it does end up complaining about Flash being disabled even if it runs in a browser where we'd use the native API anyway.. *Will fix*!).

**Comment by groovenectar** [comment]
Just want to note that if we did the abominable browser sniffing (like
Github) in 20 lines of code this would be done by now and perfectly
reliable in this case

**Comment by hallvors** [comment]
@groovenectar well, in this case we have feature detection in approx two lines of code and we haven't yet found a browser where it's not reliable.

(Regarding time taken - to begin with I underestimated the complexity of the whole project and thought it would be easy to split the whole thing into two separate scripts - one for Flash and one for native. It would certainly be sweet to ensure browsers with native support didn't have to load and parse all the Flash-related code, right? Turns out it wasn't so simple and once I started testing pages with several ZC instances requiring different behaviours I realised my first approach was a bad idea and started over. If you think you have a better approach just code it and send a competing pull request, the more the merrier.)

**Comment by hallvors** [comment]
BTW - here's the fix for the "emits Flash errors even if it intends to use native API" bug: https://github.com/zeroclipboard/zeroclipboard/pull/650/commits/26a3aa656e9093e90f7532272055ede3b3b078c8
Thanks, @peteallen .

**Comment by teppeis** [comment]
FYI: Safari 10 supports `execCommand('copy')` and disables Flash by default.
https://developer.apple.com/library/prerelease/content/releasenotes/General/WhatsNewInSafari/Articles/Safari_10_0.html
https://webkit.org/blog/6589/next-steps-for-legacy-plug-ins/

**Comment by tomsommer** [comment]
Can we get this committed?

**Comment by hallvors** [comment]
I think it's pretty ready to be committed - but I'm afraid it's flagged as decreasing test coverage somewhat. I'm not entirely sure how to fix that (and I no longer work for the employee who was funding this stuff so far, so I'll have less time available to figure things out..).

**Comment by hallvors** [comment]
Maybe @JamesMGreene has thoughts on fixing the test coverage..

**Comment by groovenectar** [comment]
Is there anything specific that needs more testing now?

**Comment by tomsommer** [comment]
Chrome is blocking Flash now, please merge this.

**Comment by squarefractal** [comment]
Is there any news on the progress of this?

**Comment by groovenectar** [comment]
I'm not sure what the issue is currently, but could throw this in there to get it working quickly and reliably:

https://github.com/zeroclipboard/zeroclipboard/issues/171#issuecomment-148152109

It would return an id of the browser if there's any need to know it, or false if the browser doesn't support native clipboard... I used this on the application side to determine whether to inject ZeroClipboard or use native and it worked well

**Comment by hallvors** [comment]
State: The PR https://github.com/zeroclipboard/zeroclipboard/pull/650 is readyish - I just merged in master and fixed some conflicts a few weeks ago but apparently later changes on master caused some more conflicts that need resolving..

**Comment by JamesMGreene** [comment]
Just so everyone is clear on the state of things here, as I know everyone is dying to get this long overdue feature: I have 1 more bug (#533) that I'd like to fix before releasing `2.3.0`, though I will bump the bug into a later release if I can't hammer it out in the next 2-3 days. After `2.3.0` goes out, I intend to immediately deal with @hallvors's PR #650 and get this feature (and only this feature) released in `2.4.0` ASAP. Thanks for all your patience! :heart:

**Comment by mc0** [comment]
@JamesMGreene is there a path forward for this? I'm not meaning to be impatient, just curious what the direction is of this feature as of April.

**Comment by krisztianb** [comment]
Hi guys. I found this page via Google search looking for a way to feature detect W3C Clipboard API support in the browser. Have you found a solution for this?

**Comment by cpeterso** [comment]
@krisztianb, I recommend checking out the [clipboard.js](https://clipboardjs.com/) library.
